src/blob.jl

```diff
@@ -119,23 +119,46 @@ end
 
 # Recursion scales better than splatting for large numbers of fields.
 Base.@assume_effects :foldable @inline function blob_offset(::Type{T}, i::Int) where {T}
-    _recursive_sum_field_sizes(T, Val(i - 1))
+    blob_offsets(T)[i]
 end
 
+Base.@assume_effects :foldable @inline function blob_offsets(::Type{T}) where {T}
+    _recursive_field_offsets(T)
+end
+Base.@assume_effects :foldable _recursive_field_offsets(::Type{T}) where {T} =
+    _recursive_field_offsets(T, Val(fieldcount(T)))
+Base.@assume_effects :foldable _recursive_field_offsets(::Type, ::Val{0}) = ()
+Base.@assume_effects :foldable _recursive_field_offsets(::Type, ::Val{1}) = (0,)
+Base.@assume_effects :foldable function _recursive_field_offsets(::Type{T}, ::Val{i}) where {T,i}
+    tup = _recursive_field_offsets(T, Val(i-1))
+    return (tup..., tup[end] + self_size(fieldtype(T, i-1)))
+end
+
+
 # Manually write a compile-time loop in the type domain, to enforce constant-folding the
-# fieldidx even for large structs (with e.g. 100 fields). This might make compiling a touch
-# slower, but it allows this to work for even large structs, like the manually-written
+# fieldindexes even for large structs (with e.g. 100 fields). This might make compiling a
+# touch slower, but it allows this to work for even large structs, like the manually-written
 # `@generated` functions did before.
-@inline function fieldidx(::Type{T}, ::Val{field}) where {T,field}
-    return _fieldidx_lookup(T, Val(field), Val(fieldcount(T)))
+@inline function fieldindexes(::Type{T}) where {T}
+    return _recursive_fieldindexes(T, Val(fieldcount(T)))
+end
+_recursive_fieldindexes(::Type{T}, ::Val{0}) where {T} = ()
+function _recursive_fieldindexes(::Type{T}, ::Val{i}) where {T,i}
+    next = _recursive_fieldindexes(T, Val(i-1))
+    names = (fieldnames(typeof(next))..., fieldname(T, i))
+    return NamedTuple{names}((next..., i))
```

Comment:
Similar question here regarding annotations :).

Reply:
oh huh, looks like it works without them here...

I guess yeah the annotations probably aren't needed because the `Val` forces everything to be operating in the compiler domain anyway. 🤷 

I dunno. I like the idea of having the annotations to force the folding? But i'm open to whatever. Do you have a preference? here and above?